linalg/hiop.hpp

```diff
@@ -0,0 +1,197 @@
+﻿// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_HIOP
+#define MFEM_HIOP
+
+#include "linalg.hpp"
+#include "../config/config.hpp"
+#include "../general/globals.hpp"
+
+#ifdef MFEM_USE_MPI
+#include "operator.hpp"
+#endif
+
+#ifdef MFEM_USE_HIOP
+
+#include "hiopInterface.hpp"
+#include "hiopNlpFormulation.hpp"
+
+namespace mfem
+{
+
+/// Adapts the OptimizationProblem class to HiOp's interface.
+class HiopOptimizationProblem : public hiop::hiopInterfaceDenseConstraints
+{
+private:
+
+#ifdef MFEM_USE_MPI
+   MPI_Comm comm_;
+#endif
+
+   // Problem info.
+   const OptimizationProblem &problem;
+
+   // Local and global number of variables and constraints.
+   const long long ntdofs_loc, m_total;
+   long long ntdofs_glob;
+
+   // Initial guess.
+   const Vector *x_start;
+
+   Vector constr_vals;
+   DenseMatrix constr_grads;
+   bool constr_info_is_current;
+   void UpdateConstrValsGrads(const Vector x);
+
+public:
+   HiopOptimizationProblem(const OptimizationProblem &prob)
+      : problem(prob),
+        ntdofs_loc(prob.input_size), ntdofs_glob(ntdofs_loc),
+        m_total(prob.GetNumConstraints()),
+        x_start(NULL),
+        constr_vals(m_total), constr_grads(m_total, ntdofs_loc),
+        constr_info_is_current(false)
+  { 
+#ifdef MFEM_USE_MPI
+    // Used when HiOp with MPI support is called by a serial driver.
+    comm_ = MPI_COMM_WORLD;
+#endif
+  }
+
+#ifdef MFEM_USE_MPI
+   HiopOptimizationProblem(const MPI_Comm& _comm,
+                           const OptimizationProblem &prob)
+      : comm_(_comm),
+        problem(prob),
+        ntdofs_loc(prob.input_size), ntdofs_glob(0),
+        m_total(prob.GetNumConstraints()),
+        x_start(NULL),
+        constr_vals(m_total), constr_grads(m_total, ntdofs_loc),
+        constr_info_is_current(false)
+   {
+      MPI_Allreduce(&ntdofs_loc, &ntdofs_glob, 1, MPI_LONG_LONG_INT,
+                    MPI_SUM, comm_);
+   }
+#endif
+
+   void setStartingPoint(const Vector &x0) { x_start = &x0; }
+
+   /** Extraction of problem dimensions:
+    *  n is the number of variables, m is the number of constraints. */
+   virtual bool get_prob_sizes(long long int& n, long long int& m);
```

Comment:
The methods in this class have a different spelling convention than most MFEM classes. Typically we do `WordOtherWord`, not `word_other_word`. I don't personally care but maybe someone will.

Reply:
This class has to implement the methods of hiop::hiopInterfaceDenseConstraints.

Reply:
I'll move it to mfem::internal to denote it shouldn't be used by users.